transport/tlsfrag/writer.go

```diff
@@ -0,0 +1,177 @@
+func newClientHelloFragWriter(base io.Writer, frag FragFunc) (io.Writer, error) {
+	if base == nil {
+		return nil, errors.New("base writer must not be nil")
+	}
+	if frag == nil {
+		return nil, errors.New("frag callback function must not be nil")
+	}
+	fw := &clientHelloFragWriter{
+		base:     base,
+		frag:     frag,
+		helloBuf: newClientHelloBuffer(),
+	}
+	if rf, ok := base.(io.ReaderFrom); ok {
+		return &clientHelloFragReaderFrom{fw, rf}, nil
+	}
+	return fw, nil
+}
+
+// Write implements io.Writer.Write. It attempts to split the data received in the first one or more Write call(s)
+// into two TLS records if the data corresponds to a TLS Client Hello record.
+func (w *clientHelloFragWriter) Write(p []byte) (n int, err error) {
+	if !w.done {
+		// not yet splitted, append to the buffer
+		if w.record == nil {
+			if n, err = w.helloBuf.Write(p); err == nil {
+				// all written, but Client Hello is not fully received yet
+				return
+			}
+			p = p[n:]
+			if errors.Is(err, errTLSClientHelloFullyReceived) {
+				w.splitHelloBufToRecord()
+			} else {
+				w.copyHelloBufToRecord()
+			}
+		}
+		// already splitted (but previous Writes might fail), try to flush all remaining w.record to w.base
+		if _, err = w.flushRecord(); err != nil {
+			return
+		}
+	}
+
+	if len(p) > 0 {
+		m, e := w.base.Write(p)
+		n += m
+		err = e
+	}
+	return
+}
+
+// ReadFrom implements io.ReaderFrom.ReadFrom. It attempts to split the first packet into two TLS records if the data
+// corresponds to a TLS Client Hello record. And then copies the remaining data from r to the base io.Writer until EOF
+// or error.
+//
+// If the first packet is not a valid TLS Client Hello, everything from r gets copied to the base io.Writer as is.
+//
+// It returns the number of bytes read. Any error except EOF encountered during the read is also returned.
+func (w *clientHelloFragReaderFrom) ReadFrom(r io.Reader) (n int64, err error) {
+	if !w.done {
+		// not yet splitted, append to the buffer
+		if w.record == nil {
+			if n, err = w.helloBuf.ReadFrom(r); err == nil {
+				// EOF, but Client Hello is not fully received yet
+				return
+			}
+			if errors.Is(err, errTLSClientHelloFullyReceived) {
+				w.splitHelloBufToRecord()
+			} else {
+				w.copyHelloBufToRecord()
+			}
+		}
+		// already splitted (but previous Writes might fail), try to flush all remaining w.record to w.base
+		if _, err = w.flushRecord(); err != nil {
+			return
+		}
+	}
+
+	m, e := w.baseRF.ReadFrom(r)
+	n += m
+	err = e
+	return
+}
+
+// copyHelloBufToRecord copies w.helloBuf into w.record without allocations.
+func (w *clientHelloFragWriter) copyHelloBufToRecord() {
+	w.record = bytes.NewBuffer(w.helloBuf.Bytes())
+	w.helloBuf = nil // allows the GC to recycle the memory
+}
+
+// splitHelloBufToRecord splits w.helloBuf into two records and put them into w.record without allocations.
+func (w *clientHelloFragWriter) splitHelloBufToRecord() {
+	received := w.helloBuf.Bytes()
```

Comment:
There are some ambiguous names here. Perhaps call them originalRecord, headRecord and tailRecord to be clearer?:
```suggestion
	originalRecord := w.helloBuf.Bytes()
```